src/server/admin/server/convert.go

```diff
@@ -0,0 +1,341 @@
+		return nil
+	}
+	return &pps.Egress{
+		URL: e.URL,
+	}
+}
+
+func convert1_7ResourceSpec(r *pps_1_7.ResourceSpec) *pps.ResourceSpec {
+	if r == nil {
+		return nil
+	}
+	return &pps.ResourceSpec{
+		Cpu:    r.Cpu,
+		Memory: r.Memory,
+		Gpu:    r.Gpu,
+		Disk:   r.Disk,
+	}
+}
+
+func convert1_7Input(i *pps_1_7.Input) *pps.Input {
+	if i == nil {
+		return nil
+	}
+	convert1_7AtomInput := func(i *pps_1_7.AtomInput) *pps.PFSInput {
+		if i == nil {
+			return nil
+		}
+		return &pps.PFSInput{
+			Name:       i.Name,
+			Repo:       i.Repo,
+			Branch:     i.Branch,
+			Commit:     i.Commit,
+			Glob:       i.Glob,
+			Lazy:       i.Lazy,
+			EmptyFiles: i.EmptyFiles,
+		}
+	}
+	convert1_7CronInput := func(i *pps_1_7.CronInput) *pps.CronInput {
+		if i == nil {
+			return nil
+		}
+		return &pps.CronInput{
+			Name:   i.Name,
+			Repo:   i.Repo,
+			Commit: i.Commit,
+			Spec:   i.Spec,
+			Start:  i.Start,
+		}
+	}
+	convert1_7GitInput := func(i *pps_1_7.GitInput) *pps.GitInput {
+		if i == nil {
+			return nil
+		}
+		return &pps.GitInput{
+			Name:   i.Name,
+			URL:    i.URL,
+			Branch: i.Branch,
+			Commit: i.Commit,
+		}
+	}
+	return &pps.Input{
+		// Note: this is deprecated and replaced by `PfsInput`
+		Pfs:   convert1_7AtomInput(i.Atom),
+		Cross: convert1_7Inputs(i.Cross),
+		Union: convert1_7Inputs(i.Union),
+		Cron:  convert1_7CronInput(i.Cron),
+		Git:   convert1_7GitInput(i.Git),
+	}
+}
+
+func convert1_7Inputs(inputs []*pps_1_7.Input) []*pps.Input {
+	if inputs == nil {
+		return nil
+	}
+	result := make([]*pps.Input, 0, len(inputs))
+	for _, i := range inputs {
+		result = append(result, convert1_7Input(i))
+	}
+	return result
+}
+
+func convert1_7Service(s *pps_1_7.Service) *pps.Service {
+	if s == nil {
+		return nil
+	}
+	return &pps.Service{
+		InternalPort: s.InternalPort,
+		ExternalPort: s.ExternalPort,
+		IP:           s.IP,
+	}
+}
+
+func convert1_7ChunkSpec(c *pps_1_7.ChunkSpec) *pps.ChunkSpec {
+	if c == nil {
+		return nil
+	}
+	return &pps.ChunkSpec{
+		Number:    c.Number,
+		SizeBytes: c.SizeBytes,
+	}
+}
+
+func convert1_7SchedulingSpec(s *pps_1_7.SchedulingSpec) *pps.SchedulingSpec {
+	if s == nil {
+		return nil
+	}
+	return &pps.SchedulingSpec{
+		NodeSelector:      s.NodeSelector,
+		PriorityClassName: s.PriorityClassName,
+	}
+}
+
+func default1_7HashtreeRoot(s string) string {
+	if s == "/" || s == "." {
+		return ""
+	}
+	return s
+}
+
+// clean canonicalizes 'path' for a Pachyderm 1.7 hashtree
+func clean1_7HashtreePath(p string) string {
+	if !strings.HasPrefix(p, "/") {
+		p = "/" + p
+	}
+	return default1_7HashtreeRoot(pathlib.Clean(p))
+}
+
+func convert1_7HashTree(storageRoot string, old *hashtree_1_7.HashTreeProto) (hashtree.HashTree, error) {
+	t, err := hashtree.NewDBHashTree(storageRoot)
+	if err != nil {
+		return nil, fmt.Errorf("could not create converted hashtree: %v", err)
+	}
+	var convert func(path string) error // forward-declare b/c recursive
+	convert = func(path string) error {
```

Comment:
This is a little wonky, couldn't it just pass the `hashtree` as an argument and be a top level function?

Reply:
Done.